library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,212 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//!
+//! # Strict Provenance
+//!
+//! This section is *non-normative* and describes an experimental set of APIs that help tools
+//! that validate the memory-safety of your program's execution. Notably this includes [miri][]
+//! and [CHERI][], which can detect when you access out of bounds memory or otherwise violate
+//! Rust's memory model.
```

Comment:
It might be worth mentioning that provenance *does* exist normatively, but that its exact shape is yet to be determined. The *reason* Miri has provenance is that it *definitely* exists as a concept in the spec. It's not just something Miri does for the fun of it.

Reply:
Yes, we probably want to have separately noted "provenance" and "experimental model: strict provenance".

I think Rust is tilted fairly strongly to a "provenance not via integers" model, so we can assume things like "integers don't have provenance, pointers do", and that everything else is about how to mechanize that.

I will try to draft some text on this.

Reply:
> I think Rust is tilted fairly strongly to a "provenance not via integers" model,

I certainly am. :)

However I am also not sure if I have yet convinced @joshtriplett that Rust definitely has provenance in the sense of an explicit spec ingredient that is 'real' in the Abstract Machine (and alters program behavior!) even when it is not real in the final executable.

Reply:
When last I saw Josh explain his model, my understanding was that he believed provenance exists and that integers don't carry it (thus 2 + 2 is subject to CSE and reassociation and all that) but that provenance has "out-of-line" accounting, akin to a (possibly more limited/restricted form of) the PNVI-ae* models.

Reply:
So the reason I left all discussion of provenance *inside* the non-normative portion is because I would like to be able to land these APIs without actually entering a discussion of the *actual* semantics of provenance are and in what ways this is an extension of the "core" principles.

Reply:
I'll carry on further expounding on provenance into another PR since it's Actually Unrelated, and if it becomes destined to land after this PR I will just add a commit uplifting text as appropriate.

Reply:
That said I think the non-normative section should properly note something to the effect of "references induce additional temporal rules that aren't described here, *something something* borrow stack is a thing *something something* not a problem this project is specifically trying to deal with"

Reply:
(Possibly totally wrong terminology abuse here, have been trying to read more papers and catch up on some of this stuff.)

With apologies to Ralf who I'm sure has a more coherent vision that unifies everything, I am kinda cleaving the memory model into its "spatial" and "temporal" components with this experiment. What I describe here is largely *only* the spatial components of the memory model, leaving all the temporal components as "stacked borrows". (The one exception here is appealing to actual deallocation, which is a temporal property but like the "free square" of memory model bingo cards.)

There are two reasons I find focusing on the spatial component appealing to focus on:

* CHERI as I understand it mostly focuses on taming spatial validity, so getting a solid spatial story is materially useful for physically running compiling programs to that target.
* Stacked borrows as I understand it mostly focuses on taming temporal validity. Like yes it has the spatial rules embedded in it but the most interesting and hairy problems it tackles seem to be the temporal ones, and it seems like it needs "extra help" from the larger ecosystem to tame the spatial component?

Reply:
I don't think I quite agree that this can be split in 'spatial' and 'temporal' (Stacked Borrows certainly has aspects of both). However, what you call Strict Provenance is certainly a good warm-up for Stacked Borrows aka "even stricter provenance". Basically, where you describe provenance with exactly one "unique identifier" per allocation, with Stacked Borrows we actually have many identifiers per allocation and some rules that govern which ones are 'active' and allowed to be used when.

So, I would not say they are orthogonal. Rather, Stacked Borrows further "refines" this model of provenance and introduces extra distinctions *within* an allocation. Once we have Stacked Borrows, we don't actually need any other kind of provenance any more.

Reply:
I 100% agree with your description of Stacked Borrows as a layer on top of "Strict Provenance". So are you comfortable with this text including something to that effect?

Reply:
Specifically something to the effect of:

Provenance is both spatial (pointers have access to specific ranges) and temporal (pointers may become invalidated). This experiment is mostly focused on trying to enforce stricter correctness in the spatial component. It is further refined by *Stacked Borrows* which enforces stricter correctness in the temporal component. Stacked Borrows are necessary to properly described the semantics of references and reborrowing, because borrows are inherently temporal.

Reply:
As I said I wouldn't tie this to spatial/temporal. Invalidating provenance on deallocation (e.g. the odd rule for ZST references) is clearly temporal. Stacked Borrows also enforces spatial bounds.

So I would say something more like
> This experiment is mostly focused on trying to enforce stricter correctness on the whole-allocation level, i.e., confiding the use of a pointer (and everything derived from it) to the bounds of an allocated object. It is further refined by Stacked Borrows which enforces stricter correctness by making further distinctions *within* an allocation. Stacked Borrows are necessary to properly described the semantics of references and reborrowing, where even pointers to the same allocated object need to be distinguished (and thus carry different provenance).